File: python/42.Trapping Rain Water.py
# class Solution:
#     def trap(self, height):
#         """
#         :type height: List[int]
#         :rtype: int
#         """
#         dic = {}    #高度 =》下标
#         block = 0
#         for i in range(len(height)):
#             h = height[i]
#             for j in range(1, h+1):
#                 pos = dic.get(j, -1)
#                 if pos != -1:
#                     block += (i - pos - 1)
#                 dic[j] = i
#         return block

# if __name__ == '__main__':
# 	print(Solution().trap([1]))
# 上述解法超时

class Solution:
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        if len(height) < 2:
            return 0
        max, maxIndex = -1, 0
        for i in range(len(height)):
            if (height[i]) > max:
                max = height[i]
                maxIndex = i

        block = 0
        leftHeight, rightHeight = height[0], height[len(height) -1]

        for i in range(maxIndex):
            if height[i] > leftHeight:
                leftHeight = height[i]
            else:
                block += leftHeight - height[i]

        for i in range(len(height)-1, maxIndex, -1):
            if height[i] > rightHeight:
                rightHeight = height[i]
            else:
                block += rightHeight - height[i]

        return block

if __name__ == "__main__":
    print(Solution().trap([0,1,0,2,1,0,1,3,2,1,2,1]))

File: cpp/42.Trapping Rain Water.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 2)
            return 0;
        int max_height = -1;
        int max_index =  -1;
        for (int i=0; i<height.size(); i++){
            if (height[i] > max_height){
                max_index = i;
                max_height = height[i];
            }
        }

        int trap_size = 0;

        int left_height = height[0];
        int right_height = height[height.size()-1];
        for (int i=1; i< max_index; i++){
            if (height[i] < left_height)
                trap_size += left_height - height[i];
            else
                left_height = height[i];
        }
        for (int i=height.size()-2; i>max_index; i--){
            if (height[i] < right_height)
                trap_size += right_height - height[i];
            else
                right_height = height[i];
        }

        return trap_size;
    }
};
